### DISCUSSION TITLE
How to Secure GRE tunnel traffic with Strongswan

### DISCUSSION THREAD

**Discussion by ssai12** [strongswan/strongswan#1243] [Q&A]
I have two rockylinux(8.6) servers running docker network either side. Using IPtables and Firewall, I could able to create a GRE tunnel b/n the two servers successfully and could also able to ping docker IPs from either sides. As we know GRE as a protocol only encapsulates packets, but doesn't provide any encryption.

Now, I have a challenge to encrypt the GRE tunnel traffic with Strongswan. I have done some background to understand Strongswan, there are usecases for creating a tunnel, but my use-case is specific to only Encrypting the GRE traffic.

I would appreciate, Any quick help. Many thanks.

**Comment by Thermi** [comment]
use %dynamic[gre] for local_ts and remote_ts. Only the [gre] part is important. You can add it to anything. Works with protocols and ports, too. E.g. [tcp/80], [udp/53], [sctp/5252].

**Reply by ssai12** [reply]
Hello @Thermi.. Thanks for quick response.
I was actually trying out with ipsec.conf/ipsec.secrets in combination with certs.
Is there a way to accomplish with ipsec.conf. I see local & remote ts are part of swanctl conf and this corresponds to left&right subnet in ipsec.conf.

`
# _**ipsec.conf**_ - strongSwan IPsec configuration file

# basic configuration

config setup
        charondebug="ike 1, knl 1, cfg 0, net 1"
        strictcrlpolicy=no
        uniqueids=yes
        cachecrls=no

conn %default
        ikelifetime=60m
        keylife=20m
        rekeymargin=3m
        keyingtries=1
       authby=secret
        keyexchange=ikev2
        mobike=no

conn host-host
     left=10.127.5.141
     leftsubnet=192.168.1.0/24
     leftid=dc1
     leftfirewall=yes
     forceencaps=yes
     right=10.127.5.245
     rightsubnet=192.168.2.0/24
     rightid=dc2
     auto=route
     ike=aes128-sha1-modp768!
     esp=aes256-sha1!
     type=tunnel
     #leftcert=dc1.pem
     leftauth=psk
     rightauth=psk

_**ipsec.secret**_

# ipsec.secrets - strongSwan IPsec secrets file

10.127.1.2 : PSK "P@ssw0rd"
`

This is my ipsec conf and secret configuration. With this configuration, when I restart strongswan, I don't see any connection (0 for both up and connecting)

am trying to understand, how to get the things working for me.. but I'm missing something in my understanding. Could you pls help? Many thanks in Advance

**Reply by ssai12** [reply]
Hi @Thermi,

I started  experimenting with  swanctl conf and below is my configuration.

**_On DC1_**
```
  connections {
    net-net {
      remote_addrs = 10.127.5.2
      local {
        auth = psk
      }
      remote {
        auth = psk
        id = "DC2"
      }
      children {
        net-net {
          local_ts  = 192.168.1.0/24[gre]
          remote_ts = 192.168.2.0/24[gre]
          start_action = trap
        }
      }
    }
  }
  
secrets {
        # loaded automatically from /etc/swanctl/private/

        psk-dc1 {
                id=dc1
                secret="P@ssw0rd" 
        }

}
```
**_On  DC2_**

```
  connections {
    net-net {
      remote_addrs = 10.127.5.1
      local {
        auth = psk
      }
      remote {
        auth = psk
        id = "DC1"
      }
      children {
        net-net {
          local_ts  = 192.168.2.0/24[gre]
          remote_ts = 192.168.1.0/24[gre]
          start_action = trap
        }
      }
    }
  }
  
secrets {
        # loaded automatically from /etc/swanctl/private/

        psk-dc2 {
                id=dc2
                secret="P@ssw0rd" 
        }

}
```
Unlike strongswan, where we use restart for reload configuration. Am not sure, but after looking at documentation, I started using 'swanctl --load-conns' for restart the configuration (not sure anything else is needed, pls correct me).
Then I see one connection successful.

- How to verify whether my swanctl conf is loaded and is it applied.
- where can I find the logs (I don't see any syslog, message folder under /var/log). I was looking Journatclt but couldn't understand much. I see 
Aug 21 11:18:17 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb activated
Aug 21 11:18:18 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb deactivated
Aug 21 11:18:18 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb deleted

- Third things, if my above conf is correctly applied and things are working as expected, then I want to experiment by giving wrong secret, so that data over the tunnel doesn't pass (because authentication will fail)

**Comment by Thermi** [comment]
Remove the networks, strongswan doesn't know about gre. That's completely a kernel thing. Strongswan in this regard (gre in IPsec) just negotiates the SPs and SAs.
You need to set up the gre stuff itself (interfaces, routing of traffic into the gre Interface) yourself.

Am 21. August 2022 11:29:43 UTC schrieb ssai12 ***@***.***>:
>Hi @Thermi,
>
>I started  experimenting with  swanctl conf and below is my configuration.
>
>**_On DC1_**
>```
>  connections {
>    net-net {
>      remote_addrs = 10.127.5.233
>      local {
>        auth = psk
>      }
>      remote {
>        auth = psk
>        id = "DC2"
>      }
>      children {
>        net-net {
>          local_ts  = 192.168.1.0/24[gre]
>          remote_ts = 192.168.2.0/24[gre]
>          start_action = trap
>        }
>      }
>    }
>  }
>  
>secrets {
>        # loaded automatically from /etc/swanctl/private/
>
>        psk-dc1 {
>                id=dc1
>                secret="netvanta" 
>        }
>
>}
>```
>**_On  DC2_**
>
>```
>  connections {
>    net-net {
>      remote_addrs = 10.127.5.131
>      local {
>        auth = psk
>      }
>      remote {
>        auth = psk
>        id = "DC1"
>      }
>      children {
>        net-net {
>          local_ts  = 192.168.2.0/24[gre]
>          remote_ts = 192.168.1.0/24[gre]
>          start_action = trap
>        }
>      }
>    }
>  }
>  
>secrets {
>        # loaded automatically from /etc/swanctl/private/
>
>        psk-dc2 {
>                id=dc2
>                secret="netvanta" 
>        }
>
>}
>```
>Unlike strongswan, where we use restart for reload configuration. Am not sure, but after looking at documentation, I started using 'swanctl --load-conns' for restart the configuration (not sure anything else is needed, pls correct me).
>Then I see one connection successful.
>
>- How to verify whether my swanctl conf is loaded and is it applied.
>- where can I find the logs (I don't see any syslog, message folder under /var/log). I was looking Journatclt but couldn't understand much. I see 
>Aug 21 11:18:17 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb activated
>Aug 21 11:18:18 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb deactivated
>Aug 21 11:18:18 host-10-127-5-233 charon-systemd[93537]: interface veth858a3fb deleted
>
>- Third things, if my above conf is correctly applied and things are working as expected, then I want to experiment by giving wrong secret, so that data over the tunnel doesn't pass (because authentication will fail)
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/1243#discussioncomment-3440557
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from mobile. Please excuse any typos or brevity

**Reply by ssai12** [reply]
Hello @Thermi,
I don't think I correctly understood your response above. I have created GRE tunnel b/n rockylinux server manually with IPtables and firewall commands. Earlier we used ipsec-tools package to secure GRE tunnel. Since, RHEL 8 ipsec-tools  is not supported and Strongswan is the alternative. I wanted to secure GRE tunnel and traffic with strongswan. **Is it possible to only do IPSEC of GRE with strongswan**?

After lot of headbanging's, I could able to get a working swanctl configuration.

```
connections {
    version = 2
    net-net {
      local_addrs = 10.127.5.233
      remote_addrs = 10.127.5.131
      local {
        auth = psk
        id = DC1
      }
      remote {
        auth = psk
      }
      children {
        snet-snet {
          local_ts  = 192.168.2.0/24[gre]
          remote_ts = 192.168.1.0/24[gre]
          start_action = trap
        }
      }
    }
  }

secrets {
        ike-DC1 {
                id=DC1
                secret = "abcdef"
        }
}
```
Now that I got some working configuration, now this configuration should stop data transfer over the GRE on authentication failure. For e.g. Suppose if I force a wrong secret on peer and fail authentication. Is this possible to do? if so how can I achieve it.

Thanks for the help, I am fighting this issue from long time and running out of time. So looking for a conclusion.

**Reply by Thermi** [reply]
use `swanctl` with corresponding arguments (run as root always, or via sudo) to control the daemon (except restarting, stopping it).
Check what configs are loaded with `swanctl -L`, load all config files, secrets, ... from swanctl related files and directories with `swanctl -q`. Check current SAs with `swanctl -l` (small L). Policies with -P, initiating SAs with -i, terminating them with -t.

Make sure automatic configuration loading is configured in strongswan.conf or related included config files (/etc/strongswan.d/charon.conf, /etc/strongswan.d/charon-custom.conf (you'd need to create that first. I keep my settings in there), ...) via charon.start-scripts or via the systemd unit ExecStartPost (either-or).

Example:
```
charon {
    start-scripts {
        load-config = /bin/swanctl -q --noprompt
    }
}
```

Key names within start-scripts are arbitrary.

The config files within /etc/strongswan.d are included by /etc/strongswan.conf via an include with a glob.

Considering you're on Rocky, the config files are probably all in /etc/strongswan/. So basically instead of /etc/strongswan.d it's /etc/strongswan/strongswan.d, instead of /etc/swanctl.conf it's /etc/strongswan/swanctl.conf, or instead of /etc/swanctl/pubkey it's /etc/strongswan/swanctl/pubkey. (just all examples here)#

> For e.g. Suppose if I force a wrong secret on peer and fail authentication. Is this possible to do? if so how can I achieve it.

Yes, just configure a wrong one.

**Reply by ssai12** [reply]
> Yes, just configure a wrong one.
I reason behind testing this fail authentication scenario is, why authentication fails, data shouldn't be replicated from dc's through GRE. Because of wrong secret.

I tried this, so the sequence of steps performed are,

- On DC1

1. After writing swanctl.conf, ran swanctl -q and then swanctl -i --child <reference-name> --ike <reference-name>

- On DC2

1. Just got the right swanctl.conf and ran swanctl -q and then swanctl --install

Scenario #1 
With matching secret key.. I could able to see data getting replicated from Dc1-Dc2

Scenario #2
I changed the secret in swanctl.conf in DC2, and then reloaded the conns using swanctl -q

Now, since the connection is already running as I didn't re-initiate connection on DC1, will the re-authentication happen automatically?

If I manually re-initiate on DC1 it would fail as expected because of secret being different.

And, when the secret is changed anytime in the middle, it should re-authenticate and when failed.. Data shouldn't be replicated.

**Reply by Thermi** [reply]
> Now, since the connection is already running as I didn't re-initiate connection on DC1, will the re-authentication happen automatically?

AFAIR reauthentication is disabled by default because of issues with continuing to forward traffic during reauthentication.

You can force reauthentication via `swanctl --rekey --reauth`.

> And, when the secret is changed anytime in the middle, it should re-authenticate and when failed.. Data shouldn't be replicated.

strongSwan doesn't do that. Secrets are only exchanged/used during authentication. Not during the use of an already eastblished SA. You can just take down a tunnel with `swanctl -t` and then try to initiate it again. With the wrong secrets that should fail. You can then also check that the tunnel fails into a closed state.

**Reply by ssai12** [reply]
As I said above, I created a GRE tunnel with IPtables and firewalld commands, Even before strongswan and stuff. I could ping, docker subnet behind the IPs from either DCs.
Strongswan comes in to picture now, to secure the GRE traffic. if the change the secret on peer2 and re-initiate the connection on peer1. It simply auth_fails.
But, GRE tunnel which is not created by strongswan will still be running and I can still ping and data will be replicated.

**Reply by ssai12** [reply]
Is there any chance I can talk to you and may be better explain my use-case?

**Reply by Thermi** [reply]
> But, GRE tunnel which is not created by strongswan will still be running and I can still ping and data will be replicated.

With start_action=trap it should be blackholed implicitely. Can you check the policies with `swanctl -P`?

You could also make an explicit drop policy by specifying the same TS in a second connection's children section but with mode=drop. That should definitely help.

Other than that you can drop the packets in *filter OUTPUT with -p gre -m policy --pol none --dir out -j DROP

**Reply by ssai12** [reply]
output from

> swanctl -p
net-net/snet-snet, TUNNEL
  local:  192.168.1.0/24[gre]
  remote: 192.168.2.0/24[gre]

Can you pls give an example for this, as I am new to strongswan.. finding it difficult to get it.

> You could also make an explicit drop policy by specifying the same TS in a second connection's children section but with mode=drop. That should definitely help.
> 
> Other than that you can drop the packets in *filter OUTPUT with -p gre -m policy --pol none --dir out -j DROP

my swanctl.conf configuration
```
  connections {
    version  = 2
    net-net {
      local_addrs = 10.127.5.131
      remote_addrs = 10.127.5.233
      local {
        auth = psk
        id = DC1
      }
      remote {
        auth = psk
      }
      children {
        snet-snet {
          local_ts  = 0.0.0.0/0[gre]
          remote_ts = 0.0.0.0/0[gre]
          start_action = trap
        }
      }
    }
  }

secrets {
        # loaded automatically from /etc/swanctl/private/

        ike-DC1 {
                id=DC1
                secret = "netvanta"
        }
}

include conf.d/*.conf
```

**Comment by Thermi** [comment]
Like a passthrough policy, just with mode=drop

Check https://docs.strongswan.org/docs/5.9/config/quickstart.html#_passthroughbypass_policies

Am 21. August 2022 18:21:22 UTC schrieb ssai12 ***@***.***>:
>output from
>
>> swanctl -p
>net-net/snet-snet, TUNNEL
>  local:  192.168.1.0/24[gre]
>  remote: 192.168.2.0/24[gre]
>
>Can you pls give an explain for this, as I am new to strongswan.. finding it difficult to get it.
>
>> You could also make an explicit drop policy by specifying the same TS in a second connection's children section but with mode=drop. That should definitely help.
>> 
>> Other than that you can drop the packets in *filter OUTPUT with -p gre -m policy --pol none --dir out -j DROP
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/1243#discussioncomment-3441965
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from mobile. Please excuse any typos or brevity
